File: Medium/2265. Count Nodes Equal to Average of Subtree/2265. Count Nodes Equal to Average of Subtree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def averageOfSubtree(self, root: Optional[TreeNode]) -> int:
        result = [0]  # Using a list to store the result to make it mutable

        def traverse(node):
            if not node:
                return (0, 0)

            left_sum, left_count = traverse(node.left)
            right_sum, right_count = traverse(node.right)

            curr_sum = node.val + left_sum + right_sum
            curr_count = 1 + left_count + right_count

            if curr_sum // curr_count == node.val:
                result[0] += 1

            return (curr_sum, curr_count)

        traverse(root)

        return result[0]

File: Medium/2265. Count Nodes Equal to Average of Subtree/2265. Count Nodes Equal to Average of Subtree.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::rc::Rc;
impl Solution {
    pub fn average_of_subtree(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        fn traverse(node: Option<Rc<RefCell<TreeNode>>>, result: &mut i32) -> (i32, i32) {
            if let Some(n) = node {
                let n = n.borrow();
                let (left_sum, left_count) = traverse(n.left.clone(), result);
                let (right_sum, right_count) = traverse(n.right.clone(), result);

                let curr_sum = n.val + left_sum + right_sum;
                let curr_count = 1 + left_count + right_count;

                if curr_sum / curr_count == n.val {
                    *result += 1;
                }

                (curr_sum, curr_count)
            } else {
                (0, 0)
            }
        }

        let mut result = 0;
        traverse(root, &mut result);
        result
    }
}
